packages/nextjs/src/server/clerkMiddleware.ts

```diff
@@ -0,0 +1,187 @@
+import type { AuthenticateRequestOptions, AuthObject, ClerkRequest, RequestState } from '@clerk/backend/internal';
+import { AuthStatus, constants, createClerkRequest, redirect } from '@clerk/backend/internal';
+import type { NextMiddleware } from 'next/server';
+import { NextResponse } from 'next/server';
+
+import { appendDevBrowserOnCrossOrigin, isRedirect, setHeader } from '../utils';
+import { clerkClient } from './clerkClient';
+import { PUBLISHABLE_KEY, SECRET_KEY } from './constants';
+import type { AuthProtect } from './protect';
+import { createProtect } from './protect';
+import type { NextMiddlewareEvtParam, NextMiddlewareRequestParam, NextMiddlewareReturn } from './types';
+import {
+  decorateRequest,
+  decorateResponseWithObservabilityHeaders,
+  handleMultiDomainAndProxy,
+  setRequestHeadersOnNextResponse,
+} from './utils';
+
+const PROTECT_REWRITE = 'CLERK_PROTECT_REWRITE';
+const PROTECT_REDIRECT_TO_URL = 'CLERK_PROTECT_REDIRECT_TO_URL';
+const PROTECT_REDIRECT_TO_SIGN_IN = 'CLERK_PROTECT_REDIRECT_TO_SIGN_IN';
+
+type ClerkMiddlewareAuthObject = AuthObject & {
+  protect: AuthProtect;
+  redirectToSignIn: (opts?: { returnBackUrl?: URL | string | null }) => Response;
+};
+
+type ClerkMiddlewareHandler = (
+  auth: () => ClerkMiddlewareAuthObject,
+  request: NextMiddlewareRequestParam,
+  event: NextMiddlewareEvtParam,
+) => NextMiddlewareReturn;
+
+type ClerkMiddlewareOptions = AuthenticateRequestOptions & { debug?: boolean };
+
+interface ClerkMiddleware {
+  /**
+   * @example
+   * export default clerkMiddleware((auth, request, event) => { ... }, options);
+   */
+  (handler: ClerkMiddlewareHandler, options?: ClerkMiddlewareOptions): NextMiddleware;
+  /**
+   * @example
+   * export default clerkMiddleware(options);
+   */
+  (options?: ClerkMiddlewareOptions): NextMiddleware;
+  /**
+   * @example
+   * export default clerkMiddleware;
+   */
+  (request: NextMiddlewareRequestParam, event: NextMiddlewareEvtParam): NextMiddlewareReturn;
+}
+
+export const clerkMiddleware: ClerkMiddleware = (...args: unknown[]): any => {
+  const [request, event] = parseRequestAndEvent(args);
```

Comment:
❓ Should we rename those 2 since there is a naming conflict with the nextMiddleware params defined below? Or this intentional? 
Reading the code inside the `clerkMiddleware` function may be confusing as to where do we intend to use  the request and event.

Reply:
Moved `const [request, event] = parseRequestAndEvent(args);` below `nextMiddleware`, thanks!